src/modules/ackermann_drive/AckermannDriveGuidance/AckermannDriveGuidance.cpp

```diff
@@ -0,0 +1,272 @@
+			float desired_heading = 0.f;
+			float lookahead_distance = math::constrain<float>(_param_ad_lookahead_tun.get() * _desired_speed,
+						   _param_ad_lookahead_min.get(), _param_ad_lookahead_max.get());
+
+			if (crosstrack_error.length() > lookahead_distance) { // Crosstrack error larger than lookahead
+				if (crosstrack_error.length() < _param_ad_lookahead_max.get()) {
+					lookahead_distance = 1.1f * crosstrack_error.length(); // Scale lookahead radius
+					desired_heading = calcDesiredHeading(_curr_wp_local, _prev_wp_local, _curr_pos_local, lookahead_distance);
+
+				} else { // Excessively large crosstrack error
+					desired_heading = calcDesiredHeading(_curr_wp_local, _curr_pos_local, _curr_pos_local, lookahead_distance);
+				}
+
+			} else { // Crosstrack error smaller than lookahead
+				desired_heading = calcDesiredHeading(_curr_wp_local, _prev_wp_local, _curr_pos_local, lookahead_distance);
+			}
+
+			// Calculate desired steering to reach lookahead point
+			float heading_error = matrix::wrap_pi(desired_heading - _vehicle_yaw);
+			_ackermann_drive_setpoint.lookahead_distance = lookahead_distance; // For logging
+			_ackermann_drive_setpoint.heading_error = (heading_error * 180.f) / (2 * M_PI_2_F); // For logging
+
+			if (math::abs_t(heading_error) <= M_PI_2_F) {
+				_desired_steering = atanf(2 * _param_ad_wheel_base.get() * sinf(heading_error) / lookahead_distance);
+
+			} else if (heading_error > 0.f) {
+				_desired_steering = atanf(2 * _param_ad_wheel_base.get() * (1.0f + sinf(heading_error - M_PI_2_F)) /
+							  lookahead_distance);
+
+			} else {
+				_desired_steering = atanf(2 * _param_ad_wheel_base.get() * (-1.0f + sinf(heading_error + M_PI_2_F)) /
+							  lookahead_distance);
+			}
+
+			_desired_steering = math::constrain(_desired_steering, -_param_ad_max_steer_angle.get(),
+							    _param_ad_max_steer_angle.get());
+
+			break;
+		}
+
+	case GuidanceState::GOAL_REACHED:
+		// Stop the rover
+		_desired_steering = 0.f;
+		_desired_speed = 0.f;
+		break;
+	}
+
+	// Publish ackermann drive setpoint
+	_ackermann_drive_setpoint.speed = _desired_speed;
+	_ackermann_drive_setpoint.steering = _desired_steering;
+	_ackermann_drive_setpoint.manual = false;
+	_ackermann_drive_setpoint.timestamp = hrt_absolute_time();
+	_ackermann_drive_setpoint_pub.publish(_ackermann_drive_setpoint);
+}
+
+float AckermannDriveGuidance::calcDesiredHeading(Vector2f const &curr_wp_local, Vector2f const &prev_wp_local,
+		Vector2f const &curr_pos_local,
+		float const &lookahead_distance)
+{
+	// Setup variables
+	float line_segment_slope = (curr_wp_local(1) - prev_wp_local(1)) / (curr_wp_local(0) - prev_wp_local(0));
+	float line_segment_rover_offset = prev_wp_local(1) - curr_pos_local(1) + line_segment_slope * (curr_pos_local(
+			0) - prev_wp_local(0));
+	float a = -line_segment_slope;
+	float c = -line_segment_rover_offset;
+	float r = lookahead_distance;
+	float EPS = 1.0E-3;
```

Comment:
Do you really need a special epsilon here? Otherwise you could use `FLT_EPSILON`